### PR TITLE
poc: load charts through IntersectionObserver

### PR DESCRIPTION
**Pull request by mlbrgl**
Loading charts and explorers through `IntersectionObserver`.

WIP, untested on many scenarios (especially dynamically added charts).

### CODE REVIEW

**Comment by danielgavrilov** [comment]
Just dumping some thoughts:

- For dynamically added charts there is a method to call on `MultiEmbedder` I think, or at least there used to be.
- I have had a painful time getting IntersectionObserver to work just right. Does it work for deep links, without any scrolling?
- There's an old issue about the "ideal way to load charts": https://github.com/owid/owid-grapher/issues/257 Basically, the interactive should not be shown until all dependencies are loaded, so that we avoid showing a loading animation completely. That would be a much smoother experience. Practically, it would mean that we request and wait for the data to load before letting React take over to render the interactive chart. I'm just sharing this in case it sparks any ideas. Definitely much harder to do.

**Comment by danielgavrilov** [comment]
Let me know if you need a full code/behaviour review on this.

**Comment by mlbrgl** [comment]
> For dynamically added charts there is a method to call on MultiEmbedder I think, or at least there used to be.

yep, I updated AdditionalInformation and RelatedCharts blocks

> I have had a painful time getting IntersectionObserver to work just right. Does it work for deep links, without any scrolling?

this time, I've kept the configuration to a minimum and have been able to make it work reliably including deep links on page load without scrolling. It does fail in [Cypress tests](https://github.com/owid/owid-grapher/blob/perf-loading-charts/devTools/cypress/integration/multiEmbedder.js) and [after rotating on mobile](https://www.notion.so/5bc73c7515084adf9e9f09762157ace6#b7df6f6578ed4423a25d23da76839e7c). 

> There's an old issue about the "ideal way to load charts"

Thanks for the background info, I'll keep that one for next time 🙂

**Comment by mlbrgl** [comment]
@danielgavrilov yes I think that'd be good, thanks for offering

Here is the scope of the changes:
- Reduce preload distance
- Reduce or remove the use of scroll events in favour of `IntersectionObserver`
- (abandoned) Prevent rendering charts if scroll by quickly

Background, rationale and preview URL: https://www.notion.so/5bc73c7515084adf9e9f09762157ace6

File: grapher/core/GrapherConstants.ts [hunk 1]
```diff
@@ -13,6 +13,8 @@ export enum ChartTypeName {
 
 export const GRAPHER_EMBEDDED_FIGURE_ATTR = "data-grapher-src"
 
+export const GRAPHER_RENDER_TIMEOUT_ID = "data-render-timeout-id"
+
```
Comment (danielgavrilov) [inline grapher/core/GrapherConstants.ts:17]: This is now unused and can be removed?

File: site/multiembedder/MultiEmbedder.tsx [hunk 2]
```diff
@@ -148,88 +49,122 @@ const globalEntitySelectorElement = () =>
     document.querySelector(GLOBAL_ENTITY_SELECTOR_ELEMENT)
 
 class MultiEmbedder {
-    private figures: EmbeddedFigure[] = []
+    // private figures: EmbeddedFigure[] = []
```
Comment (danielgavrilov) [inline site/multiembedder/MultiEmbedder.tsx:52]: can get rid of this comment?

**Review by danielgavrilov — APPROVED** [review]
Looks good, and much cleaner!

**Comment by danielgavrilov** [comment]
On the scroll performance issue: if we can't get away with waiting-for-scroll-to-stop-before-loading-charts, then it seems the options are:

- Optimizing the data parsing/transformation (specifically the table transforms are quite slow)
- Running the data parsing/transformation in a worker, not the main thread

**Pull request merged**